### PR TITLE
Added hamming, dual, and dual rail codes

### PR DESCRIPTION
**Pull request by dhruvdevu**

### CODE REVIEW

File: codes/classical/bits/hadamard.yml [hunk 1]
```diff
@@ -0,0 +1,16 @@
+code_id: hadamard
```
Comment (valbert4) [inline codes/classical/bits/hadamard.yml:1]: Spreadsheet says "simplex" is dual of Hamming. According to wikipedia, this is dual to extended Hamming..? Check Concise Encyclopedia of Coding Theory Def. 1.10.2 and Table 12.1.

File: codes/classical/bits/hamming.yml [hunk 2]
```diff
@@ -0,0 +1,40 @@
...
+#######################################################
+
+code_id: hamming
+physical: bits
+logical: bits
+
+name: 'Hamming code'
+#introduced: '\cite{doi: 10.1002/j.1538-7305.1950.tb00463.x.}'
+
+description: 'Hamming Codes are a family of linear codes with parameters \((2^r-1,2^r-r-1, 3)\) for any
```
Comment (valbert4) [inline codes/classical/bits/hamming.yml:13]: Part about error detection should be moved to protection field and removed here.

File: codes/classical/bits/hamming.yml [hunk 3]
```diff
@@ -0,0 +1,40 @@
...
+  - 'Has the maximum rate among binary codes for that block length and with minimum distance 3.
+  Therefore, it is a perfect code, i.e, it saturates the Hamming Bound.'
+
+relations:
+  parents:
+    - code_id: 'binary_linear'
+      detail: ''
+
+  cousins:
+    - code_id: hadamard
```
Comment (valbert4) [inline codes/classical/bits/hamming.yml:35]: See above, might not be correct

File: codes/classical/bits/hamming.yml [hunk 4]
```diff
@@ -0,0 +1,40 @@
...
+notes:
+  - 'Has the maximum rate among binary codes for that block length and with minimum distance 3.
+  Therefore, it is a perfect code, i.e, it saturates the Hamming Bound.'
+
+relations:
+  parents:
+    - code_id: 'binary_linear'
+      detail: ''
+
+  cousins:
```
Comment (valbert4) [inline codes/classical/bits/hamming.yml:34]: Add cousin code_id: perfect with detail: ''

File: codes/classical/q-ary_digits/dual.yml [hunk 5]
```diff
@@ -0,0 +1,20 @@
...
+The dual code swaps the generator and parity check matrices from the original code. Therefore, the dual of an \([n,k,d] \) code
+is an \([n,n-k,d'']\) code, where \(d''\) is not always related to \(d\).'
+
+notes:
+  - 'The dual of the dual code is the original code.'
+  - 'A code that is equal to its dual is a self-dual code.'
+  - 'A code that is contained in its dual is a weakly self-dual code.'
+
+relations:
+  parents:
```
Comment (valbert4) [inline codes/classical/q-ary_digits/dual.yml:18]: Two cousins: self_dual and weakly_self_dual with appropriate detail statements.

File: codes/quantum/oscillators/dual_rail.yml [hunk 6]
```diff
@@ -0,0 +1,53 @@
...
+while the logical \(\ket{1}\) state is represented by \(\ket{10}\). The original
+paper that introduced this encoding seeks to implement Deutsch''s algorithm specifically,
+i.e., it does not provide a framework for universal quantum computation (although this can
+be realized using Kerr non-linearities along with beamsplitters), and it does not solve the
+general problem of error correction.
+In the original paper
+that introduced this encoding, error correction is performed by simply rejecting
+executions which result in error states (i.e, illegal results), and re-executing.'
+
+protection: 'This code cannot protect against general errors.
```
Comment (valbert4) [inline codes/quantum/oscillators/dual_rail.yml:21]: Either explain the Caldeira-Leggett coupling and ref to errors mixing modes in the context Deutsch's problem, or omit one or both. Simply put, this code detects one loss error for each mode, which maps one of the logical codewords into the vacuum state \(|00\rangle\).

File: codes/quantum/oscillators/dual_rail.yml [hunk 7]
```diff
@@ -0,0 +1,53 @@
+code_id: dual_rail
+
+physical: qubits
+logical: qubits
+
+name: 'Dual Rail'
+
+introduced: \cite{doi:10.1103/PhysRevA.52.3489}
+
+description: 'In a dual rail qubit encoding, single qubits are represented by Fock states
```
Comment (valbert4) [inline codes/quantum/oscillators/dual_rail.yml:10]: Just keep first two sentences.

File: codes/quantum/oscillators/dual_rail.yml [hunk 8]
```diff
@@ -0,0 +1,53 @@
...
+in the wrong modes.'
+
+features:
+  general_gates:
+    - 'General gates are performed using beamsplitters and Kerr non-linearities.
+    Universal quantum computing can be achieved with photons in dual rail encodings using the
+    KLM protocol with only linear optical elements and photon detectors.'
+
+notes:
+  - 'For Deutsch''s problem with optical qubits, error correction using photon number detection
```
Comment (valbert4) [inline codes/quantum/oscillators/dual_rail.yml:36]: - For first note, add ref to Deutsch's problem.
- Move second note to cousin with code_id: chuang-leung-yamamoto
- Merge last note into protection section; I think it is basically equivalent to saying this is an error detecting code against loss.

File: codes/classical/bits/hamming.yml [hunk 9]
```diff
@@ -0,0 +1,40 @@
...
+## This is a code entry in the error correction zoo. ##
+##       https://github.com/errorcorrectionzoo       ##
+#######################################################
+
+code_id: hamming
+physical: bits
+logical: bits
+
+name: 'Hamming code'
+#introduced: '\cite{doi: 10.1002/j.1538-7305.1950.tb00463.x.}'
```
Comment (valbert4) [inline codes/classical/bits/hamming.yml:11]: Is there a ref for this?

**Review by valbert4 — CHANGES_REQUESTED** [review]

File: codes/classical/bits/hamming.yml (hunk 9 above)
Comment (dhruvdevu) [reply codes/classical/bits/hamming.yml]: Fixed issue with doi.

File: codes/quantum/oscillators/dual_rail.yml (hunk 8 above)
Comment (dhruvdevu) [reply codes/quantum/oscillators/dual_rail.yml]: fixed

File: codes/quantum/oscillators/dual_rail.yml (hunk 7 above)
Comment (dhruvdevu) [reply codes/quantum/oscillators/dual_rail.yml:10]: fixed

File: codes/classical/bits/hamming.yml [hunk 10]
```diff
@@ -0,0 +1,40 @@
...
+
+features:
+  rate: '\(1-\frac{\log n}{n}\)'
+
+realizations:
+  - 'Commonly used when error rates are very low, for example, computer RAM.'
+
+notes:
+  - 'Has the maximum rate among binary codes for that block length and with minimum distance 3.
+  Therefore, it is a perfect code, i.e, it saturates the Hamming Bound.'
```
[the remark this answers is not in the document]
Comment (dhruvdevu) [reply codes/classical/bits/hamming.yml:27]: I believe it's not the only one, for example - https://en.wikipedia.org/wiki/Binary_Golay_code

File: codes/classical/bits/hamming.yml (hunk 3 above)
Comment (dhruvdevu) [reply codes/classical/bits/hamming.yml:35]: Wiki says the dual of hamming is the shortened hadamard code.. should I correct this to shortened hadamard, and change the hadamard entry to 'extended hamming'?

**Comment by dhruvdevu** [comment]
I've responded to most of the changes and also added a page for fusion based quantum computing (in commit 3430041)

File: codes/classical/bits/hamming.yml (hunk 3 above)
Comment (valbert4) [reply codes/classical/bits/hamming.yml:35]: Keep hadamard as cousin. Add detail entry saying the dual of hamming is shortened version of hadamard.

**Comment by valbert4** [comment]
Thanks @dhruvdevu. Could you please submit a fresh commit with all your code files? I can't seem to merge the FBQC commit you made.

**Pull request closed without merging**